### ISSUE TITLE
Stack leak in recursive functions with bind

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#468]
Recursive functions with a binder end up generating a new `FUnionEnv` frame for every recursive call.  Memory usage thus grows linearly with the number of recursive calls, and it can takes a long time to unwind all the `FUnionEnv` frames once the recursion completes.

I noticed this when I programmed robots to wait until seeing a certain entity on their cell.  After they had been waiting for a while, it took them a long time to respond once the correct entity was placed; in fact, their response time seemed proportional to the amount of time they had been waiting.

**Comment by byorgey** [comment]
If you want to reproduce this, try the following.  Start a new game in creative mode, and load the definitions:
```
def fmap : (a -> b) -> cmd a -> cmd b = \f. \x.
  a <- x; return (f a)
end

def ifC : cmd bool -> {cmd a} -> {cmd a} -> cmd a = \test. \then. \else.
  b <- test; if b then else
end

def while : cmd bool -> cmd a -> cmd () = \test. \body.
  ifC test {body ; while test body} {}
end

def until = \test. while (fmap not test) end
def waitUntil = \test. until test noop end
```
Then run the command
```
build {move; waitUntil (ishere "tree"); move}
```
Now run
```
build {move; create "tree"; place "tree"}
```
Repeat this experiment twice.  The first time, run the second `build` immediately after the first.  You should see the first robot `move` after just a small delay.  The second time, wait 5 minutes or so after running the first `build` before running the second. This time, you should observe a few seconds' delay (I observe about 6 seconds of delay) before the first robot moves.

**Comment by byorgey** [comment]
Just for interest's sake, I initially titled this issue "tail call optimization" since it seemed to have something to do with recursion and I thought of the kind of growing call stack that would be dealt with via tail call optimization.  However, when I looked into it and thought about it a bit more, I realized that using a CEK machine essentially does TCO automatically, "for free".  The stack does not record function calls, but rather the continuation, i.e. context.  When we make a recursive call in tail position, the context --- *i.e.* what we expect to do with the result --- has not changed, so the stack does not grow.
